runtime/dart_isolate.cc

```diff
@@ -0,0 +1,615 @@
+  );
+
+  if (!service_isolate) {
+    *error = strdup("Could not create the service isolate.");
+    return nullptr;
+  }
+
+  // The engine never holds a strong reference to the VM service isolate. Since
+  // we are about to lose our last weak reference to it, start the VM service
+  // while we have this reference.
+  const auto& settings = vm->GetSettings();
+
+  const bool running_from_sources =
+      !DartVM::IsRunningPrecompiledCode() && vm->GetPlatformKernel() == nullptr;
+
+  tonic::DartState::Scope scope(service_isolate.get());
+  if (!DartServiceIsolate::Startup(
+          settings.ipv6 ? "::1" : "127.0.0.1",  // server IP address
+          settings.observatory_port,            // server observatory port
+          tonic::DartState::HandleLibraryTag,   // embedder library tag handler
+          running_from_sources,                 // running from source code
+          false,  //  disable websocket origin check
+          error   // error (out)
+          )) {
+    // Error is populated by call to startup.
+    return nullptr;
+  }
+
+  return service_isolate->isolate();
+}
+
+// |Dart_IsolateCreateCallback|
+Dart_Isolate DartIsolate::DartIsolateCreateCallback(
+    const char* advisory_script_uri,
+    const char* advisory_script_entrypoint,
+    const char* package_root,
+    const char* package_config,
+    Dart_IsolateFlags* flags,
+    void* callback_data,
+    char** error) {
+  if (callback_data == nullptr &&
+      strcmp(advisory_script_uri, "vm-service") == 0) {
+    // The VM attempts to start the VM service for us on |Dart_Initialize|. In
+    // such a case, the callback data will be null and the script URI will be
+    // "vm-service". In such cases, we just create the service isolate like
+    // normal but dont hold a reference to it at all. We also start this isolate
+    // since we will never again reference it from the engine.
+    return DartCreateAndStartServiceIsolate(advisory_script_uri,         //
+                                            advisory_script_entrypoint,  //
+                                            package_root,                //
+                                            package_config,              //
+                                            flags,                       //
+                                            callback_data,               //
+                                            error                        //
+    );
+  }
+
+  return CreateDartVMAndEmbedderObjectPair(
+             advisory_script_uri, advisory_script_entrypoint, package_root,
+             package_config, flags, callback_data, error)
+      .first;
+}
+
+std::pair<Dart_Isolate, fxl::WeakPtr<DartIsolate>>
+DartIsolate::CreateDartVMAndEmbedderObjectPair(
+    const char* advisory_script_uri,
+    const char* advisory_script_entrypoint,
+    const char* package_root,
+    const char* package_config,
+    Dart_IsolateFlags* flags,
+    void* callback_data,
```

Comment:
I'd prefer declaring this argument as a DartIsolate*.
Callers who receive it from the Dart VM as a void* can cast it to signify that they know what is in the callback data.

Reply:
Good idea. I used `DartIsolate*` as the argument everywhere and cast it once when I gave these as the callbacks to use to the VM in `Dart_Initialize`.